src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -4027,22 +4094,45 @@ def get_or_create_run(
         Args:
             pipeline_run: The pipeline run to get or create.
 
+        Raises:
+            ValueError: If the request
         Returns:
             The pipeline run, and a boolean indicating whether the run was
             created or not.
         """
+        if not pipeline_run.orchestrator_run_id:
+            raise ValueError(
+                "Unable to get or create run for request with missing "
+                "orchestrator run ID."
+            )
+
+        try:
+            return (
+                self._replace_placeholder_run(pipeline_run=pipeline_run),
+                True,
+            )
+        except KeyError:
+            # No placeholder run available for this run -> We now try to create
+            # a new run in the database, which will fail if a run with the same
+            # orchestrator_run_id and deployment_id already exists. In that case
+            # we fetch and return the existing run.
+            pass
+
         # We want to have the 'create' statement in the try block since running
         # it first will reduce concurrency issues.
         try:
             return self.create_run(pipeline_run), True
```

Comment:
If I understand it right, creation from non-placeholder run will only be reached for scheduled runs, right? In that case, you might want to make the docstrings and comments a bit clearer to explain in which scenarios which case is reached, this is currently not easy to follow

Reply:
Not only scheduled runs, but also rerunning a run on e.g. Kubeflow reuses the deployment and therefore creates new runs. I'll try to add some more comments to explain it better.